webrtc.html

```diff
@@ -11591,7 +11591,76 @@ <h3>DTMF Example</h3>
   console.log('DTMF function not available');
 }
       </pre>
-    </section><!--
+    </section>
+    <section>
+      <h3>Perfect Negotiation Example</h3>
+      <p>Perfect negotiation is a recommended pattern to abstract away signaling
+      from the rest of an application. It lets applications operate on both
+      peer connection objects at once, on either end of the call, without having
+      to worry about signaling state, racing, or glare.</p>
+      <p>It designates different roles to the two peers, with behavior to
+      resolve signaling collisions between them:</p>
+      <ol>
+        <li>
+          <p>The <dfn>polite peer</dfn> uses rollback to avoid collision with an
+          incoming offer</p>
+        </li>
+        <li>
+          <p>The <dfn>impolite peer</dfn> ignores an incoming offer when this
+          would collide with its own</p>
+        </li>
+      </ol>
+      <p>Together, they manage signaling for the rest of the application in a
+      manner that doesn't deadlock.</p>
+      <pre class="example highlight">
+// The perfect negotiation logic, separated from the rest of an application ---
+
+pc.onnegotiationneeded = async () => {
+  try {
+    await pc.setLocalDescription();
+    signaling.send({description: pc.localDescription});
+  } catch (e) {
+     console.log(e);
+  }
+};
+
+let offerIgnored = false;
+
+signaling.onmessage = async ({data: {description, candidate}}) => {
+  try {
+     if (description) {
+      if (!polite && description.type == "offer" && pc.signalingState != "stable") {
+        offerIgnored = true;
+        return;
+      }
+      offerIgnored = false;
+      await pc.setRemoteDescription(description); // SRD rolls back as needed
+      if (description.type == "offer") {
+        await pc.setLocalDescription();
+        signaling.send({description: pc.localDescription});
+      }
+    } else if (candidate) {
+      try {
+        await pc.addIceCandidate(candidate);
+      } catch (e) {
+        if (!offerIgnored) throw e; // Suppress ignored offer's candidates
```

Comment:
Are we guaranteed that if we ignore a candidate, the equivalent candidate will be re-generated at the other endpoint and sent again later so that we always end up getting remote candidates? Or is it possible that somehow candidates get lost and we never connect in some weird edge cases?

Reply:
@henbos Note we're not ignoring candidates, only errors from them bouncing.